### PR TITLE
[Core] use std filesystem

### PR DESCRIPTION
**Pull request by philbucher**
**📝 Description**
Since we now use C++17, we can finally directly use `std::filesystem`
I hence removed the auxiliry library that we were using and started to use std::filesystem directly. To keep this PR small, I only did some small refactorings, more will come in future PRs

**🆕 Changelog**
- use `std::filesystem` instead of `ghc::filesystem`

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
I recommend to define an auxiliary file with the following include: (C++17 for filesystem is not uniformally implemented)

~~~c++
// System includes
// We haven't checked which filesystem to include yet
#ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
// Check for feature test macro for <filesystem>
#   if defined(__cpp_lib_filesystem)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
// Check for feature test macro for <experimental/filesystem>
#   elif defined(__cpp_lib_experimental_filesystem)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
// We can't check if headers exist...
// Let's assume experimental to be safe
#   elif !defined(__has_include)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
// Check if the header "<filesystem>" exists
#   elif __has_include(<filesystem>)
// If we're compiling on Visual Studio and are not compiling with C++17, we need to use experimental
#     ifdef _MSC_VER
// Check and include header that defines "_HAS_CXX17"
#       if __has_include(<yvals_core.h>)
#         include <yvals_core.h>
// Check for enabled C++17 support
#         if defined(_HAS_CXX17) && _HAS_CXX17
// We're using C++17, so let's use the normal version
#           define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
#         endif
#       endif
// If the marco isn't defined yet, that means any of the other VS specific checks failed, so we need to use experimental
#       ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
#         define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
#       endif
// Not on Visual Studio. Let's use the normal version
#     else // #ifdef _MSC_VER
#       define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
#     endif
// Check if the header "<filesystem>" exists
#   elif __has_include(<experimental/filesystem>)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
// Fail if neither header is available with a nice error message
#   else
#     error Could not find system header "<filesystem>" or "<experimental/filesystem>"
#   endif
// We priously determined that we need the exprimental version
#   if INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
#     include <experimental/filesystem>
// We need the alias from std::experimental::filesystem to std::filesystem
namespace std {
  namespace filesystem = experimental::filesystem;
}
// We have a decent compiler and can use the normal version
#   else
#     include <filesystem>
#   endif
#endif // #ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
~~~

**Comment by loumalouomega** [comment]
> I recommend to define an auxiliary file with the following include: (C++17 for filesystem is not uniformally implemented)
> 
> ```c++
> // System includes
> // We haven't checked which filesystem to include yet
> #ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> // Check for feature test macro for <filesystem>
> #   if defined(__cpp_lib_filesystem)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
> // Check for feature test macro for <experimental/filesystem>
> #   elif defined(__cpp_lib_experimental_filesystem)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> // We can't check if headers exist...
> // Let's assume experimental to be safe
> #   elif !defined(__has_include)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> // Check if the header "<filesystem>" exists
> #   elif __has_include(<filesystem>)
> // If we're compiling on Visual Studio and are not compiling with C++17, we need to use experimental
> #     ifdef _MSC_VER
> // Check and include header that defines "_HAS_CXX17"
> #       if __has_include(<yvals_core.h>)
> #         include <yvals_core.h>
> // Check for enabled C++17 support
> #         if defined(_HAS_CXX17) && _HAS_CXX17
> // We're using C++17, so let's use the normal version
> #           define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
> #         endif
> #       endif
> // If the marco isn't defined yet, that means any of the other VS specific checks failed, so we need to use experimental
> #       ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> #         define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> #       endif
> // Not on Visual Studio. Let's use the normal version
> #     else // #ifdef _MSC_VER
> #       define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
> #     endif
> // Check if the header "<filesystem>" exists
> #   elif __has_include(<experimental/filesystem>)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> // Fail if neither header is available with a nice error message
> #   else
> #     error Could not find system header "<filesystem>" or "<experimental/filesystem>"
> #   endif
> // We priously determined that we need the exprimental version
> #   if INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> #     include <experimental/filesystem>
> // We need the alias from std::experimental::filesystem to std::filesystem
> namespace std {
>   namespace filesystem = experimental::filesystem;
> }
> // We have a decent compiler and can use the normal version
> #   else
> #     include <filesystem>
> #   endif
> #endif // #ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> ```

In my experience

File: kratos/CMakeLists.txt [hunk 1]
```diff
@@ -86,6 +86,12 @@ if(${USE_TRIANGLE_NONFREE_TPL} MATCHES ON )
   set(KRATOS_EXTERNAL_LIBRARIES ${KRATOS_EXTERNAL_LIBRARIES} triangle)
 endif(${USE_TRIANGLE_NONFREE_TPL} MATCHES ON )
 
+if(${CMAKE_COMPILER_IS_GNUCXX})
```
Comment (loumalouomega) [inline kratos/CMakeLists.txt:92]: Try to avoid this using my suggestion

Comment (philbucher) [reply kratos/CMakeLists.txt:92]: This cannot be avoided, see the end of the documentation https://en.cppreference.com/w/cpp/filesystem
GCC initially had filesystem implemented in a separate lib (until v9), hence before that it has to be linked explicitly
Hence I am quite sure your suggestion wouldnt change this

Overall I think filesystem is quite well integrated by now, see the compiler support: https://en.cppreference.com/w/cpp/compiler_support/17
Hence unless someone has problems I would leave the include as default for now

**Review by loumalouomega — APPROVED** [review]

File: kratos/CMakeLists.txt (hunk 1 above)
Comment (loumalouomega) [reply kratos/CMakeLists.txt:92]: Okay, this is my personal experience

Comment (philbucher) [reply kratos/CMakeLists.txt:92]: which systems/compilers did you have problems with?

Comment (loumalouomega) [reply kratos/CMakeLists.txt:92]: Clang in Mac and different versions of GCC

Comment (philbucher) [reply kratos/CMakeLists.txt:92]: ok good to know

Please create issues in case it causes problems

**Comment by philbucher** [comment]
Hm weird
Did you find anything related online?
Can you compile a simple program including this header?

Clang 13 seems way to new to have this issue

I think it should work without extra linking since clang 9

**Comment by loumalouomega** [comment]
@rubenzorrilla  n you try with my proposal?

~~~c++
// System includes
// We haven't checked which filesystem to include yet
#ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
// Check for feature test macro for <filesystem>
#   if defined(__cpp_lib_filesystem)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
// Check for feature test macro for <experimental/filesystem>
#   elif defined(__cpp_lib_experimental_filesystem)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
// We can't check if headers exist...
// Let's assume experimental to be safe
#   elif !defined(__has_include)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
// Check if the header "<filesystem>" exists
#   elif __has_include(<filesystem>)
// If we're compiling on Visual Studio and are not compiling with C++17, we need to use experimental
#     ifdef _MSC_VER
// Check and include header that defines "_HAS_CXX17"
#       if __has_include(<yvals_core.h>)
#         include <yvals_core.h>
// Check for enabled C++17 support
#         if defined(_HAS_CXX17) && _HAS_CXX17
// We're using C++17, so let's use the normal version
#           define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
#         endif
#       endif
// If the marco isn't defined yet, that means any of the other VS specific checks failed, so we need to use experimental
#       ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
#         define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
#       endif
// Not on Visual Studio. Let's use the normal version
#     else // #ifdef _MSC_VER
#       define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
#     endif
// Check if the header "<filesystem>" exists
#   elif __has_include(<experimental/filesystem>)
#     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
// Fail if neither header is available with a nice error message
#   else
#     error Could not find system header "<filesystem>" or "<experimental/filesystem>"
#   endif
// We priously determined that we need the exprimental version
#   if INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
#     include <experimental/filesystem>
// We need the alias from std::experimental::filesystem to std::filesystem
namespace std {
  namespace filesystem = experimental::filesystem;
}
// We have a decent compiler and can use the normal version
#   else
#     include <filesystem>
#   endif
#endif // #ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
~~~

**Comment by loumalouomega** [comment]
> @rubenzorrilla n you try with my proposal?
> 
> ```c++
> // System includes
> // We haven't checked which filesystem to include yet
> #ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> // Check for feature test macro for <filesystem>
> #   if defined(__cpp_lib_filesystem)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
> // Check for feature test macro for <experimental/filesystem>
> #   elif defined(__cpp_lib_experimental_filesystem)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> // We can't check if headers exist...
> // Let's assume experimental to be safe
> #   elif !defined(__has_include)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> // Check if the header "<filesystem>" exists
> #   elif __has_include(<filesystem>)
> // If we're compiling on Visual Studio and are not compiling with C++17, we need to use experimental
> #     ifdef _MSC_VER
> // Check and include header that defines "_HAS_CXX17"
> #       if __has_include(<yvals_core.h>)
> #         include <yvals_core.h>
> // Check for enabled C++17 support
> #         if defined(_HAS_CXX17) && _HAS_CXX17
> // We're using C++17, so let's use the normal version
> #           define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
> #         endif
> #       endif
> // If the marco isn't defined yet, that means any of the other VS specific checks failed, so we need to use experimental
> #       ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> #         define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> #       endif
> // Not on Visual Studio. Let's use the normal version
> #     else // #ifdef _MSC_VER
> #       define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 0
> #     endif
> // Check if the header "<filesystem>" exists
> #   elif __has_include(<experimental/filesystem>)
> #     define INCLUDE_STD_FILESYSTEM_EXPERIMENTAL 1
> // Fail if neither header is available with a nice error message
> #   else
> #     error Could not find system header "<filesystem>" or "<experimental/filesystem>"
> #   endif
> // We priously determined that we need the exprimental version
> #   if INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> #     include <experimental/filesystem>
> // We need the alias from std::experimental::filesystem to std::filesystem
> namespace std {
>   namespace filesystem = experimental::filesystem;
> }
> // We have a decent compiler and can use the normal version
> #   else
> #     include <filesystem>
> #   endif
> #endif // #ifndef INCLUDE_STD_FILESYSTEM_EXPERIMENTAL
> ```

If this works, we can define an auxiliary file again

**Comment by rubenzorrilla** [comment]
Just mentioning that I tried the hard way, including `<experimental/filesystem>` and substituting `std::experimental::filesystem::`. This works.

**Comment by loumalouomega** [comment]
> Just mentioning that I tried the hard way, including `<experimental/filesystem>` and substituting `std::experimental::filesystem::`. This works.

Can you try https://github.com/KratosMultiphysics/Kratos/tree/core/fix-include-filesystem ?

**Comment by philbucher** [comment]
Why the full version?
If it works with the experimental namespace I would keep it as simple as possible

**Comment by loumalouomega** [comment]
> Why the full version? If it works with the experimental namespace I would keep it as simple as possible

I can tell you from experience that in MacOS you may need the full version...

**Comment by loumalouomega** [comment]
> > Why the full version? If it works with the experimental namespace I would keep it as simple as possible
> 
> I can tell you from experience that in MacOS you may need the full version...

I will simplify

**Pull request merged**